### DISCUSSION TITLE
WithPerceptionTracking not fixing perception warnings on Stacks

### DISCUSSION THREAD

**Discussion by lucianolang** [pointfreeco/swift-composable-architecture#3518] [Q&A]
We've encountered a weird case in which wrapping a ForEach in a LazyVStack does not fix perception warnings 

This code does not produce any warnings inside this view
```
WithPerceptionTracking {
          if store.transationsState.isEmpty {
            EmptyTransactionListRow(title: L10n.Home.TransactionList.emptyState)
          } else {
            WithPerceptionTracking {
              ForEach(store.scope(state: \.transationsState, action: \._transactionList)) { itemStore in
                WithPerceptionTracking {
                  if let header = itemStore.billsDescriptor.value {
                    Section(header: BillHeader(header.billheaderDescriptor)) {
                      TransactionListView(store: itemStore)
                    }
                  } else {
                    TransactionListView(store: itemStore)
                  }
                }
              }
            }
          }
        }
```

However, including a LazyVStack breaks the tests:

```
WithPerceptionTracking {
      LazyVStack(spacing: Spacing.xSmall.rawValue) {
        WithPerceptionTracking {
          if store.transationsState.isEmpty {
            EmptyTransactionListRow(title: L10n.Home.TransactionList.emptyState)
          } else {
            WithPerceptionTracking {
              ForEach(store.scope(state: \.transationsState, action: \._transactionList)) { itemStore in
                WithPerceptionTracking {
                  if let header = itemStore.billsDescriptor.value {
                    Section(header: BillHeader(header.billheaderDescriptor)) {
                      TransactionListView(store: itemStore)
                    }
                  } else {
                    TransactionListView(store: itemStore)
                  }
                }
              }
            }
          }
        }
      }
    }
```

Not sure what were missing here, would appreciate some help...

Xcode Version 16.1
Simulator iPhone 16 Pro iOS 18.0

**Comment by mbrandonw** [comment]
Hi @lucianolang, there isn't enough information here to help. Can you please try creating a minimal project from scratch that demonstrates the behavior you are seeing and share that?

**Comment by lucianolang** [comment]
Hi @mbrandonw, thank you for your quick reply... I did what you suggested and know I'm even more confused as I have created a project with a similar structure to the one I have and found that tests successfully complete without the need for a single use of `WithPerceptionTracking`. I have read the [documentation](https://swiftpackageindex.com/pointfreeco/swift-perception/1.3.5/documentation/perception/withperceptiontracking) and added, I believe, many possible instances that could trigger the runtime warning, but still nothing.

Perhaps if I could get some light as to why the issue is not being replicated in this project I could work out a solution to the problem in my original project... 

[withPerceptionTrackingIssue.zip](https://github.com/user-attachments/files/18009095/withPerceptionTrackingIssue2.zip)

Thank you so much,

**Reply by mbrandonw** [reply]
> and found that tests successfully complete without the need for a single use of `WithPerceptionTracking`.

The project you have shared targets iOS 18. The `WithPerceptionTracking` stuff is only needed when deploying on an older iOS. In fact, if you were to use `WithPerceptionTracking` in your project you would get a deprecation warning:

> ⚠️ 'WithPerceptionTracking' was deprecated in iOS 17: 'WithPerceptionTracking' is no longer needed in iOS 17+

**Reply by lucianolang** [reply]
Oh! I overlooked this detail, apologies. So I changed the minimum deployment target to iOS 15.6 and I am still not getting the perception error. 

[withPerceptionTrackingIssue2.zip](https://github.com/user-attachments/files/18011229/withPerceptionTrackingIssue2.zip)

**Reply by mbrandonw** [reply]
Hey @lucianolang, you did uncover a small bug we have in TCA which I just PR'd a fix for #3521. However, that bug isn't really related to your core problem. It is just a small thing that affects views that use `ForEach` at the root of their view hierarchy.

So, for your demo, can you wrap your `LazyVStack` in `WithPerceptionTracking`:

```swift
var body: some View {
  WithPerceptionTracking {
    LazyVStack {
```

…in order to work around the bug, and then let us know what the problem is that you are observing?

**Reply by lucianolang** [reply]
That's great! The main issue in my original project is that no matter how many WithPerceptionTracking I use to wrap my views I still get the error:
> Perceptible state was accessed but is not being tracked. 

In this minimal project that I shared with you I dont get such error, even after changing the minimum deployment target, or adding escaping, trailing closures, LazyVStack and other elements that require "perception tracking"... 

I'm now trying to understand what could possibly be causing the issue in my original project by understanding why is it that it is not happening in the test project. 

I.E. Why am I not able to see the error "Perceptible state was accessed but is not being tracked" in the test project when I run the tests?? 

Thank you!!
[withPerceptionTrackingIssue.zip](https://github.com/user-attachments/files/18021460/withPerceptionTrackingIssue2.zip)

**Reply by lucianolang** [reply]
Hey @mbrandonw! So I managed to trigger the unctracked perceptible change by accessing the state from outside the LazyVStack... Once I had this error, wrapping the LazyVStack in WithPerceptionTracking eliminated the issue...

```
      WithPerceptionTracking {
        VStack {
          Text(store.title)
          LazyVStack {
            ForEach(store.scope(state: \.childStates, action: \.childAction)) { state in
              Text(state.id)
              Section {
                Text(state.id)
                ChildView(store: state) {
                  Text(state.id)
                }
              }
            }
          }
        }.onAppear {
          store.send(.action)
        }
```

However I tried to point my project to this branch perception-fix-identified-array and im still getting the perception error that cannot be solved by wrapping the view in a WithPerceptionTracking...
